plugins/base-gradle-plugin/src/main/groovy/org/kordamp/gradle/plugin/base/extensions/Repository.java

```diff
@@ -0,0 +1,13 @@
+package org.kordamp.gradle.plugin.base.extensions;
+
+import org.gradle.api.Named;
+import org.gradle.api.provider.Property;
+
+public interface Repository extends Named {
+
+    Property<String> getUrl();
+    // Property<Credentials> getCredentials();
```

Comment:
@melix I have some very strange behaviour here. if remove the comment here and the particular configuration here

https://github.com/kordamp/kordamp-gradle-plugins/pull/496/files#diff-7eb248e07942ac8642c2618d3e65cb667d28fc87dffa45c67d740b7a9022f214R50-R53

then actually the whole extension cannot be found causing no such property `kordamp` for the build script.

what is the best way how to provide 

```
credentials {
    username = 'foo'
    pasword = 'bar'
}
```

for objects used inside `NamedDomainObjectContainer`?

I've tired to let `Repository` implements `ExtensionAware` and made `credentials` an extension but it didn't work.

and when I tried to add default methods `credentials {}` then I was missing the value to delegate to and I would required `ObjectFactory` to create the default values. Does having `Property<Credentials>` have actually have any benefit here as the `Repository` creation should be lazy itself or can I simply use `Credentials` real object here?

Reply:
It's a bit difficult to tell without seeing the actual problems. There shouldn't be any problem with using a `Property<Credentials>`, but it's true that you should avoid using default methods because as far as I know, default methods wouldn't automatically create the `credentials { ... }` method. Wrt using `Credentials` directly, I think it depends on what you want to offer as UX. For me it makes sense to use a `Property` because you can then share the same instance with different repositories, while keeping track of how the credentials instance is created.

Reply:
and is there any simple way how to inject `PropertyFactory` into the `Repository` object if it's converted to the abstract class? I've tried it but it looks like it only accepts the name e.g. `Repository(String name) { this.name = name; }`. Or is there any better way how to get access to the `PropertyFactory`, i.g. something like `PropertyFactoryAware` interface to extend so I can create a default method

```
default void credentials(Closure closure) {
     Credentials credentials = getCredentials().convention(getPropertyFactory(}.property(new DefaultCredentials()).get();
     ConfigureUtil.configure(credentials, closure);
```

Reply:
actually I'm over engineering here, right? I can set the default value as object and I don't need it as a `Property<Credentials>` 🤦

Reply:
First, don't use `Closure`. Use `Action`, Gradle takes care of creating the `Closure` version for Groovy and the SAM receiver version for Kotlin.

Then, for creating instances in advanced use cases where you want to set the defaults for nested objects, you can take a look at how I do it in the [GraalVM native plugin](https://github.com/graalvm/native-build-tools/blob/9c8271fc3dae1b9e82d51d86cecf5ae3088da4b2/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImagePlugin.java#L304-L312). It requires a bit more work since now you'd need to provide an abstract class but then it can configure defaults: https://github.com/graalvm/native-build-tools/blob/9c8271fc3dae1b9e82d51d86cecf5ae3088da4b2/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/BaseNativeImageOptions.java#L214-L221

Reply:
thanks for the examples! ok so the solution would be to step back one level and create and configure the `NamedDomainObjectContainer<Repository>` manually while creating the extension.